apps/pkt-gen/pkt-gen.c

```diff
@@ -1140,15 +1183,19 @@ send_packets(struct netmap_ring *ring, struct pkt *pkt, void *frame,
 	for (fcnt = nfrags, sent = 0; sent < count; sent++) {
 		struct netmap_slot *slot = &ring->slot[cur];
 		char *p = NETMAP_BUF(ring, slot->buf_idx);
-		int buf_changed = slot->flags & NS_BUF_CHANGED;
+		struct pkt *old = (struct pkt*)(p - sizeof(old->vh) + vh);
+		int copy = options & OPT_COPY || slot->flags & NS_BUF_CHANGED;
+		/* sender_body() drops OPT_COPY after starting, but we need to
+		 * copy over any ARP packets lingering in the txring */
+		copy |= (old->eh.ether_type == htons(ETHERTYPE_ARP));
```

Comment:
This is likely a cache miss, I would do that only if -G is specified

Reply:
It's actually worse than it looks... Only thread 0 can have (targ->garp != 0) but here all TX threads are checking for no reason. Let me improve this.

(My initial implementation used the "sem[]" area in the netmap_ring to store a dirty flag, for when the ring needs to be checked. It works fine but it's fiddly and I don't think people would like it.)